### PR TITLE
fix(backend): change downloaded IR from JSON to YAML. Fixes #7673

### LINKED ISSUE #7673: [backend] KFPv2: Backend always return JSON format PipelineSpec template even if I upload yaml
### Environment

KFP backend 2.0.0-alpha.1

### Steps to reproduce

1. Upload a pipeline using KFPv2 yaml format
2. Get the pipeline template by calling `GET Pipeline Template` API
3. The response of template is in JSON format.

Side note: If you upload a KFPv2 json format, it returns json format template, which is expected.

### Expected result

We should preserve the format when storing pipeline template. For example, when user uploads a YAML format, it should return YAML format when calling `GET Pipeline Template` API. When user uploads a JSON format, it should return JSON format. 

/assign @Linchin 
/cc @jlyaoyuli 

---

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by Linchin**
**Description of your changes:**
Fixes #7673 
Now `GET Pipeline Template` API returns IR in YAML format.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by zijianjoy** [comment]
Previously, we return JSON pipeline template in any situation. In this PR, we return YAML pipeline template in any situation. 

Is it possible to preserve the original type we pass in to store pipeline template? For example: If I upload a json template, return a json template when I call GetTemplate API. If I upload a yaml template, return a yaml template when I call GetTemplate API.

File: backend/src/apiserver/resource/resource_manager.go [hunk 1]
```diff
@@ -1241,15 +1242,21 @@ func (r *ResourceManager) GetPipelineVersionTemplate(versionId string) ([]byte,
...
-		return nil, util.Wrap(err, "Get pipeline version template failed")
+		return nil, util.Wrap(err, "Get pipeline version template failed: cannot get file")
 	}
 
-	return template, nil
+	// convert template from JSON to YAML
+	template_yaml, err := yaml.JSONToYAML(template)
+	if err != nil {
+		return nil, util.Wrap(err, "Get pipeline version template failed: unable to convert template from JSON to YAML")
+	}
```
Comment (zijianjoy) [inline backend/src/apiserver/resource/resource_manager.go:1257]: Can we add some tests for validating this PR change?

1. If pipeline template was stored as JSON format, return a JSON format template.
2. If pipeline template was stored as YAML format, return a YAML format template.

**Comment by Linchin** [comment]
I think it would be really hard to implement this, since the backend API is not aware of which format the IR is uploaded in.

**Comment by Linchin** [comment]
This makes a lot of sense. Let me double check and test locally.

**Comment by Linchin** [comment]
Maybe we could also discuss whether we will support JSON at all? It feels like we only supported it very briefly during v2alpha0 pre-release.

**Comment by zijianjoy** [comment]
> Maybe we could also discuss whether we will support JSON at all? It feels like we only supported it very briefly during v2alpha0 pre-release.

Good point, please consult with @chensun to make a decision for this topic.

File: backend/src/apiserver/resource/resource_manager.go (hunk 1 above)
Comment (Linchin) [reply backend/src/apiserver/resource/resource_manager.go:1257]: That's a great idea! I will do it.

Comment (Linchin) [reply backend/src/apiserver/resource/resource_manager.go:1257]: That's a great idea! I will do it.

**Comment by Linchin** [comment]
Although this PR converts the stored JSON IR into YAML, it does not solve the root issue - why is the original file not stored in the first place? This conversion may lose comments that are supported for YAML files.

After some investigation, it looks like the template that is being stored in minio is marshaled back into JSON from the struct itself. As is shown in the following code:
https://github.com/kubeflow/pipelines/blob/e95b9f5f4663a0713d4e2565c686e71aa96643c1/backend/src/apiserver/resource/resource_manager.go#L289-L291

Straightforwardly, it looks like we can just change the `tmpl.Bytes()` into `pipelineFile`, which can be found among the input parameters.

**Comment by Linchin** [comment]
Although this PR converts the stored JSON IR into YAML, it does not solve the root issue - why is the original file not stored in the first place? This conversion may lose comments that are supported for YAML files.

After some investigation, it looks like the template that is being stored in minio is marshaled back into JSON from the struct itself. As is shown in the following code:
https://github.com/kubeflow/pipelines/blob/e95b9f5f4663a0713d4e2565c686e71aa96643c1/backend/src/apiserver/resource/resource_manager.go#L289-L291

Straightforwardly, it looks like we can just change the `tmpl.Bytes()` into `pipelineFile`, which can be found among the input parameters.

**Comment by zijianjoy** [comment]
> 
> Edit: after some discussion, it looks like the template was altered when it was unmarshaled, so we cannot store the original file directly. Maybe for now we will just convert the output of `tmpl.Bytes()` into yaml and only support yaml from now.

Sounds good on saving yaml directly for now. Can you also add a TODO here saying that we are losing the comments in yaml and we need to find a way to preserve the original file?

**Comment by Linchin** [comment]
Thank you @zijianjoy, I will add the TODO in the code. :)

File: backend/src/apiserver/template/v2_template.go [hunk 2]
```diff
@@ -109,10 +109,15 @@ func (t *V2Spec) Bytes() []byte {
...
+		// this is unexpected, cannot convert proto message to JSON
 		return nil
 	}
-	return bytes
+	bytesYAML, err := yaml.JSONToYAML(bytes)
+	if err != nil {
+		// this is unexpected, cannot convert JSON to YAML
+		return nil
+	}
+	return bytesYAML
```
Comment (zijianjoy) [inline backend/src/apiserver/template/v2_template.go:120]: I recommend adding a test for this change.

Comment (Linchin) [reply backend/src/apiserver/template/v2_template.go:120]: Done! Now json IR is blocked.

File: backend/src/apiserver/resource/resource_manager.go [hunk 3]
```diff
@@ -287,6 +287,8 @@ func (r *ResourceManager) CreatePipeline(name string, description string, namesp
 	}
 
 	// Store the pipeline file to a path dependent on pipeline version
+	// TODO(lingqinggan): comments of the yaml files are lost during the coversion;
```
Comment (chensun) [inline backend/src/apiserver/resource/resource_manager.go:290]: nit: of -> in

File: backend/src/apiserver/template/template_test.go [hunk 4]
```diff
@@ -73,7 +73,7 @@ kind: Workflow`,
 		templateType: V1,
 	}, {
 		template:     v2SpecHelloWorldJSON,
-		templateType: V2,
+		templateType: Unknown,
```
Comment (chensun) [inline backend/src/apiserver/template/template_test.go:76]: With this change, we lost the coverage for when templateType == V2. We should definitely have a positive case for that.

Comment (chensun) [reply backend/src/apiserver/template/template_test.go:76]: The negative (json) test case can be short and inlined here, doesn't need to be the full valid json IR.

Comment (Linchin) [reply backend/src/apiserver/template/template_test.go:76]: > With this change, we lost the coverage for when templateType == V2. We should definitely have a positive case for that.

The test file includes another test case with V2 template in yaml. So I think it is covered?



> The negative (json) test case can be short and inlined here, doesn't need to be the full valid json IR.

There is also another test case in JSON format, the content of which is not pipeline spec though. Maybe I can just remove the JSON test case (with entire pipeline)?

**Review by Arhell — COMMENTED** [review]
Please sign CLA

**Review by chensun — COMMENTED** [review]
/lgtm
/approve

Thanks!

**Pull request merged**